pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -66,12 +83,62 @@ func (t *GroupTokenBucket) Update(now time.Time) {
 		t.Tokens += float64(t.Settings.Fillrate) * delta.Seconds()
 		t.LastUpdate = &now
 	}
+	if t.Tokens >= 0 {
+		t.LoanExpireTime = nil
+	}
+	if t.Tokens > defaultMaxTokens {
+		t.Tokens = defaultMaxTokens
+	}
 }
 
-// Request requests tokens from the token bucket.
-func (t *GroupTokenBucket) Request(
+// request requests tokens from the token bucket.
+func (t *GroupTokenBucket) request(
 	neededTokens float64, targetPeriodMs uint64,
-) *rmpb.TokenBucket {
-	// TODO: Implement the token bucket algorithm.
-	return nil
+) (*rmpb.TokenBucket, int64) {
+	var res rmpb.TokenBucket
+	res.Settings = &rmpb.TokenLimitSettings{}
+	// TODO: consider the shares for dispatch the fill rate
+	res.Settings.Fillrate = 0
+	if neededTokens <= 0 {
+		return &res, 0
+	}
+
+	if t.Tokens >= neededTokens {
+		t.Tokens -= neededTokens
+		// granted the total request tokens
+		res.Tokens = neededTokens
+		return &res, 0
+	}
+
+	var grantedTokens float64
+	if t.Tokens > 0 {
+		grantedTokens = t.Tokens
+		t.Tokens = 0
+		neededTokens -= grantedTokens
+	}
+
+	var periodFilled float64
+	var trickleTime = int64(targetPeriodMs)
+	if t.LoanExpireTime != nil && t.LoanExpireTime.After(*t.LastUpdate) {
+		duration := t.LoanExpireTime.Sub(*t.LastUpdate)
+		periodFilled = float64(t.Settings.Fillrate) * (1 - loanReserveRatio) * duration.Seconds()
+		trickleTime = duration.Milliseconds()
+	} else {
+		et := t.LastUpdate.Add(t.LoanMaxPeriod)
+		t.LoanExpireTime = &et
+		periodFilled = float64(t.Settings.Fillrate) * (1 - loanReserveRatio) * t.LoanMaxPeriod.Seconds()
```

Comment:
Do we need to ensure `LoadMaxPeriod` period greater than `targetPeriodMs` ?

Reply:
Yes

Reply:
and why not `periodFilled = float64(t.Settings.Fillrate) * (1 - loanReserveRatio) * trickleTime `？

Reply:
periodFilled is the number of tokens that can be allocated in whole loan period, so I think we should use LoanMaxPeriod

Reply:
I think it will make the  tokens not be graceful consumed with `Fillrate` in the loan period. Client will consume more tokens in target period(loan expire  period> target period)

Reply:
I guess that you want the consumption curve to be smoother.  To fit this situation, we can use shorter `LoanMaxPeriod`. We can use long LoanMaxPeriod for high credit or important resource group, so they can respond more promptly to qps peaks.